### PR TITLE
Value number identities for compares and subs

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Improve value number support for various subtracts and unsigned compares.
For instance for non-float values, x - x == 0.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
FIxes some issues seen over in #15491.

Jit-diffs shows a few hits in corelib and frameworks:
```
Total bytes of diff: -681 (0.00% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
        -334 : System.Private.CoreLib.dasm (-0.01% of base)
        -127 : Microsoft.CodeAnalysis.CSharp.dasm (-0.01% of base)
         -60 : System.Memory.dasm (-0.07% of base)
         -37 : System.Data.Common.dasm (0.00% of base)
         -36 : Microsoft.CodeAnalysis.dasm (0.00% of base)
13 total files with size differences (13 improved, 0 regressed), 117 unchanged.
Top method regessions by size (bytes):
           4 : System.Web.HttpUtility.dasm - HttpEncoder:UrlEncodeUnicode(ref):ref
Top method improvements by size (bytes):
         -44 : System.Private.CoreLib.dasm - StreamWriter:Write(ref):this (2 methods)
         -37 : System.Private.CoreLib.dasm - Task:WhenAll(ref):ref (4 methods)
         -28 : System.Data.Common.dasm - SqlString:CompareBinary2(struct,struct):int
         -22 : Microsoft.CodeAnalysis.CSharp.dasm - Symbol:ValidateAttributeUsageAndDecodeWellKnownAttributes(struct,struct,struct,ref,short):ref:this
         -22 : System.Private.CoreLib.dasm - StreamWriter:Write(ref,int,int):this
47 total methods with size differences (46 improved, 1 regressed), 145517 unchanged.
```
Diff for example from #15491:
```diff
-; Lcl frame size = 40
+; Lcl frame size = 8
 
 G_M7513_IG01:
-       4883EC28             sub      rsp, 40
+       50                   push     rax
 
 G_M7513_IG02:
        488B02               mov      rax, bword ptr [rdx]
        8B5208               mov      edx, dword ptr [rdx+8]
-       448BC2               mov      r8d, edx
-       413BD0               cmp      edx, r8d
-       771B                 ja       SHORT G_M7513_IG05
-
-G_M7513_IG03:
-       418BD0               mov      edx, r8d
-       412BD0               sub      edx, r8d
-       4D63C0               movsxd   r8, r8d
-       4A8D0440             lea      rax, bword ptr [rax+2*r8]
+       4863D2               movsxd   rdx, edx
+       488D0450             lea      rax, bword ptr [rax+2*rdx]
+       33D2                 xor      edx, edx
        488901               mov      bword ptr [rcx], rax
        895108               mov      dword ptr [rcx+8], edx
        488BC1               mov      rax, rcx
 
-G_M7513_IG04:
-       4883C428             add      rsp, 40
+G_M7513_IG03:
+       4883C408             add      rsp, 8
        C3                   ret      
 
-G_M7513_IG05:
-       E88EDCF54B           call     ThrowHelper:ThrowArgumentOutOfRangeException()
-       CC                   int3     
-
-; Total bytes of code 51, prolog size 4 for method N:Slice(struct):struct
+; Total bytes of code 30, prolog size 1 for method N:Slice(struct):struct
```
Not sure why the after version still sets up a frame. I'll look into that separately.

@briansull  PTAL 
cc @dotnet/jit-contrib

**Comment by briansull** [comment]
We typically have to set up a Frame whenever the method makes calls.
We were asled to do this for ETW stack walking (uses RBP chaining)
We coudl relax this requirement when calling methods that are expected to Throw (or otherwise not return)

File: src/jit/valuenum.cpp [hunk 1]
```diff
@@ -1154,10 +1154,18 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
...
+                        // (x - x) => 0
+                        if (!varTypeIsFloating(typ))
                         {
-                            resultVN = arg0VN;
+                            ZeroVN = VNZeroForType(typ);
+                            if (arg1VN == ZeroVN)
+                            {
+                                resultVN = arg0VN;
+                            }
+                            else if ((arg0VN != NoVN) && (arg1VN == arg0VN))
```
Comment (briansull) [inline src/jit/valuenum.cpp]: We probably should add a helper method:

```
bool VNIsEqual(ValueNum vn1, ValueNum vn2)
{
   return (v1==v2)  && (vn1 != NoVN)  &&  !varTypeIsFloating(TypeOfVN(vn1);
}
```

**Comment by AndyAyersMS** [comment]
It may not be clear from the diffs -- the newly optimized version of the method no longer throws, so it is now a leaf method.

File: src/jit/valuenum.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/jit/valuenum.cpp]: Ok, will add this.

**Comment by AndyAyersMS** [comment]
Need to screen out FP in some cases rather than just assert we don't see it. Fix forthcoming.

**Comment by AndyAyersMS** [comment]
As for why there's a frame for the new version -- the ref count for V09 is too high. So while unreferenced, it appears to need a stack slot. Seems like counts get off track in constant prop. Still debugging.

**Comment by AndyAyersMS** [comment]
Similar issue to ones I have seen before -- remorphing bumps ref counts and overstates them (here if we constant prop int a statement we remorph the entire statement, so any surviving lcl vars get extra ref counts they don't deserve).

Culprit is #8284, which was necessitated to avoid undercounting ref counts for newly introduced local appearances when remorphing. Upshot is that existing references now get overcounted.

Similar issue noted in #9068.

Sample dump (with some extra output I added)
```
optVNConstPropOnTree: have constant for [000052]
N002 (  1,  1) [000051] ------------              /--*  LCL_VAR   int    V03 tmp1         u:2 <l:$200, c:$240>
N003 (  3,  3) [000052] ----G-------              *  SUB       int    $40
N001 (  1,  1) [000050] ------------              \--*  LCL_VAR   int    V03 tmp1         u:2 (last use) <l:$200, c:$240>
New refCnts for V03: refCnt =  3, refCntWtd = 6   
New refCnts for V03: refCnt =  2, refCntWtd = 4   
After constant propagation on [000052]:
     (  3,  3) [000112] ------------              *  STMT      void  (IL   ???...  ???)
N003 (  3,  3) [000052] ------------              |  /--*  CNS_INT   int    0 $40
N005 (  3,  3) [000111] -A--G---R---              \--*  ASG       int    $40
N004 (  1,  1) [000110] D------N----                 \--*  LCL_VAR   int    V09 tmp7         d:2 $40

Success, remorphing stmt [000112]
New refCnts for V09: refCnt =  3, refCntWtd = 6   
optVNAssertionPropCurStmt morphed tree:
N001 (  1,  1) [000052] ------------              /--*  CNS_INT   int    0 $40
N003 (  1,  3) [000111] -A--G---R---              *  ASG       int    $40
N002 (  1,  1) [000110] D------N----
```
V09's ref count is unaffected by the const prop, but it still gets incremented.

**Pull request merged**